resources/lua/triggers.lua

```diff
@@ -0,0 +1,203 @@
+--------------------------------------------------------------------------------
+
+local next_group_id = 1
+
+mod.TriggerGroup = {
+}
+local TriggerGroup = mod.TriggerGroup
+TriggerGroup.__index = TriggerGroup
+
+function TriggerGroup.new(id)
+    local ret = setmetatable({}, TriggerGroup)
+
+    ret.id = id
+    ret.triggers = {}
+
+    return ret
+end
+
+function TriggerGroup:add(regex_or_trigger, options, callback)
+    local trigger
+    if Trigger.isTrigger(regex_or_trigger) then
+        trigger = regex_or_trigger
+    else
+        trigger = Trigger.new(regex_or_trigger, options, callback)
+    end
+    self.triggers[trigger.id] = trigger
+    return trigger
+end
+
+function TriggerGroup:get(id)
+    return self.triggers[id]
+end
+
+function TriggerGroup:getTriggers()
+    return {table.unpack(self.triggers)}
+end
+
+function TriggerGroup:remove(id)
+    self.triggers[id] = nil
+end
+
+function TriggerGroup:clear()
+    self.triggers = {}
+end
+
+function TriggerGroup:checkLine(line)
+    local toRemove = {}
+    for _, trigger in pairs(self.triggers) do
+        trigger:checkLine(line)
+        if trigger.count == 0 then
+            toRemove[#toRemove + 1] = trigger.id
+        end
+    end
+    for _, trigger in ipairs(toRemove) do
+        self:remove(trigger)
+    end
+end
+
+--------------------------------------------------------------------------------
+-- module ----------------------------------------------------------------------
+--------------------------------------------------------------------------------
+
+mod.trigger_groups = {
+    TriggerGroup.new(1)
+}
+local user_trigger_groups = mod.trigger_groups
+
+mod.system_trigger_groups = {
+    TriggerGroup.new(1)
+}
+local system_trigger_groups = mod.system_trigger_groups
+
+local function getTriggerGroups()
+    if blight:is_core_mode() then
+        return system_trigger_groups
+    end
+    return user_trigger_groups
+end
+
+function mod.add(regex, options, callback)
+    return getTriggerGroups()[1]:add(regex, options, callback)
+end
+
+function mod.get(id)
+    for _, group in pairs(getTriggerGroups()) do
+        local trigger = group:get(id)
+        if trigger then return trigger end
+    end
+    return nil
+end
+
+function mod.getGroup(id)
+    if not id then id = 1 end
+    return getTriggerGroups()[id]
+end
+
+function mod.remove(id)
+    for _, group in pairs(getTriggerGroups()) do
+        group:remove(id)
+    end
+end
+
+function mod.clear()
+    for _, group in pairs(getTriggerGroups()) do
+        group:clear()
+    end
+end
+
+function mod.addGroup()
+    local ret = TriggerGroup.new(next_group_id)
+    getTriggerGroups()[next_group_id] = ret
+    next_group_id = next_group_id + 1
+
+    return ret
+end
+
+mud:add_output_listener(function(line)
+    for _, group in pairs(system_trigger_groups) do
+        group:checkLine(line)
+    end
+    for _, group in pairs(user_trigger_groups) do
+        group:checkLine(line)
+    end
+    return line
```

Comment:
You should differentiate between prompt and regular output here. Prompt triggers should only match prompt output. This logic is ready and available in dev now. Eg. `Line:prompt() => bool`

Reply:
Well, perhaps not here. But you get it :wink:

Reply:
Saw your comment. Ignore this